cabal-install/Distribution/Client/Manpage.hs

```diff
@@ -0,0 +1,187 @@
+{-# LANGUAGE CPP, ExistentialQuantification #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Distribution.Client.Manpage
+-- Copyright   :  (c) Maciek Makowski 2015
+-- License     :  BSD-like
+--
+-- Maintainer  :  cabal-devel@haskell.org
+-- Stability   :  provisional
+-- Portability :  non-portable (ExistentialQuantification)
+--
+-- Functions and types for building the manual page.
+
+module Distribution.Client.Manpage
+  ( -- * Command list
+    CommandVisibility (..)
+  , CommandSpec (..)
+  , commandFromSpec
+    -- * Manual page generation
+  , manpage
+  ) where
+
+import Distribution.Simple.Command
+import Distribution.Client.Setup (globalCommand)
+
+import Data.Char (toUpper)
+import Data.List (intercalate)
+
+data CommandVisibility = Visible | Hidden
+
+-- | wraps a @CommandUI@ together with a function that turns it into a @Command@.
+-- By hiding the type of flags for the UI allows construction of a list of all UIs at the
+-- top level of the program. That list can then be used for generation of manual page
+-- as well as for executing the selected command.
+data CommandSpec action
```

Comment:
This looks like something that belongs better in `D.S.Command`. Consider refactoring that module.